docs/data/base/components/number-input/number-input.md

```diff
@@ -0,0 +1,152 @@
+---
+product: base
+title: React Number Input component and hook
+components: NumberInput
+hooks: useNumberInput
+githubLabel: 'component: NumberInput'
+---
+
+# Number Input
+
+<p class="description">The Number Input component provides users with a field for integer values, and buttons to increment or decrement the value.</p>
+
+{{"component": "modules/components/ComponentLinkHeader.js", "design": false}}
+
+{{"component": "modules/components/ComponentPageTabs.js"}}
+
+## Introduction
+
+A number input is a UI element that accepts numeric values from the user.
+Base UI's Number Input component is a customizable replacement for the native HTML `<input type="number">` that solves common usability issues of the native counterpart, such as:
+
+- Inconsistencies across browsers in the appearance and behavior of the stepper buttons
+- Allowing certain non-numeric characters ('e', '+', '-', '.') and silently discarding others
+- Incompatibilities with assistive technologies and limited accessibility features
+
+See [this article](https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/) by the GOV.UK Design System team for a more detailed explanation.
+
+{{"demo": "UnstyledNumberInputIntroduction.js", "defaultCodeOpen": false, "bg": "gradient"}}
```

Comment:
I think that a denser demo could be great. Most of the number input I have ever used were small.

Current:
<img width="284" alt="Screenshot 2023-05-19 at 19 44 42" src="https://github.com/mui/material-ui/assets/3165635/803e6496-3e08-4260-a9c7-803917866678">

What it could look like:
<img width="271" alt="Screenshot 2023-05-19 at 19 44 19" src="https://github.com/mui/material-ui/assets/3165635/4eb2cf64-befe-40f7-95c0-29207712bdb4">

Reply:
Actually I think all the base demos may look better if it was more dense (closer to Joy UI) but that might be my personal preference 😅 

However if this demo was a different size than the rest of them it might stand out in an awkward way - what do you think @danilo-leal ?

Reply:
I like the experience when a component adjusts to its content— in this case, up to 3 digits. 
However, I believe that the demos would look more consistent if it follows the same padding as the default input. Hence, I think making it a bit narrower in width, and keeping the height the same, wouldn't damage the consistency. Thoughts?

Reply:
I'm down for making all of these demos a bit denser, that makes sense to me! We should probably tackle that in a separate PR so we don't bloat this one poking at other components 🤙

Reply:
Agree, let's leave it as a general improvements on the demos after the PR is merged.

Reply:
I've updated the demos to the latest denser design (e.g. https://github.com/mui/material-ui/pull/37750)